### ISSUE TITLE
ESLint doesn't handle higher-order JSX usage

### ISSUE THREAD

**Issue by aem** [eslint/eslint#6786]
**What version of ESLint are you using?**
`eslint@3.1.1`

**What parser (default, Babel-ESLint, etc.) are you using?**
`babel-eslint`

**Please show your full configuration:**
```json
{
  "parser": "babel-eslint",
  "extends": "airbnb-base",
  "plugins": ["react"],
  "rules": {
    "comma-dangle": 0,
    "max-len": 0
  }
}
```

**What did you do? Please include the actual source code causing the issue.**
I'm writing a higher-order React component (though this isn't relevant to React), and using the functions argument in JSX still triggers a `no-unused-vars` violation. The following code should be enough to reproduce:
```jsx
const HigherOrderComponent = WrappedComponent => class extends React.Component {
  /* omitted */

  render() {
    return (
      <WrappedComponent
        {...this.props}
      />
    );
  }
};

export default HigherOrderComponent;
```

**What did you expect to happen?**
Running ESLint on the above code should complete successfully with no violations

**What actually happened? Please include the actual, raw output from ESLint.**
I've modified the package name and removed references to my home directory, but all of the other output is unmodified.
```sh
$ npm run lint

> my-package@0.1.0 lint ~/git/my-package
> eslint --ext '.js' ./src


~/git/my-package/src/index.js
  51:27  error  'WrappedComponent' is defined but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

**Comment by platinumazure** [comment]
Per our [README](https://github.com/eslint/eslint/blob/master/README.md#does-eslint-support-jsx), ESLint supports the parsing of JSX but that does not give you React semantics. Look into [eslint-plugin-react](https://www.npmjs.com/package/eslint-plugin-react) and enable their `jsx-uses-vars` rule to get what you need.
